### PR TITLE
[FIX] stock: several small fixes to be suggested for review

### PR DESCRIPTION
**Pull request by wtaferner**
**Description of the issue/feature this PR addresses:**
Feel free to take a look and think about the changes...sorry for not providing clear use cases, but most of them are self explaining IMHO

**Current behavior before PR:**
Blocking use cases which were not covered by Odoo core...tests should follow...had no time as these are fixes done to move forward not to provide code quality as it should have been done in master before release ;-)

**Desired behavior after PR is merged:**
Having less issues and removing blockers...I will try to work out the tests, but I think it would be helpful to minimize this effort by just thinking about the minimal code changes first.

@amoyaux @jco-odoo @sle-odoo @pimodoo @nim-odoo 
As you guys are responsible for my pains, I invite you to review and work on this with me...I think at least one or the other should be clear for you...the others we will figure out I hope...

Merry X-Mas :santa: 

OPW-802056

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_quant.py [hunk 1]
```diff
@@ -250,9 +250,9 @@ def _update_reserved_quantity(self, product_id, location_id, quantity, lot_id=No
         rounding = product_id.uom_id.rounding
         quants = self._gather(product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict)
         available_quantity = self._get_available_quantity(product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict)
-        if float_compare(quantity, 0, precision_rounding=rounding) > 0 and float_compare(quantity, available_quantity, precision_rounding=rounding) > 0:
+        if quants and float_compare(quantity, 0, precision_rounding=rounding) > 0 and float_compare(quantity, available_quantity, precision_rounding=rounding) > 0:
             raise UserError(_('It is not possible to reserve more products of %s than you have in stock.') % (', '.join(quants.mapped('product_id').mapped('display_name'))))
-        elif float_compare(quantity, 0, precision_rounding=rounding) < 0 and float_compare(abs(quantity), sum(quants.mapped('reserved_quantity')), precision_rounding=rounding) > 0:
+        elif quants and float_compare(quantity, 0, precision_rounding=rounding) < 0 and float_compare(abs(quantity), sum(quants.mapped('reserved_quantity')), precision_rounding=rounding) > 0:
```
Comment (wtaferner) [inline addons/stock/models/stock_quant.py]: These were having issues during inventory adjustments...blocking an adjustment is a little bit odd as we should not care during this procedure or only when we actually have quants available and we have no reserved quants to return anyway without quants to return.

**Comment by sle-odoo** [comment]
Hello @wtaferner,
Merry Christmas.

I see in your branch that a few tests are broken, see [this one](https://github.com/odoo/odoo/blob/11.0/addons/stock/tests/test_quant.py#L482) that seems legit to me.

As always, it'll be nice that you provide usecases for these patches... Testcases would be better, but usecases so I can reproduce the issue and try to work on it would be enough. Can you provide at least one usecase so we can work in a constructive way on these patches?

Regards

**Comment by wtaferner** [comment]
@sle-odoo 
Hmm, you are starting with the most difficult one to argue I guess, I think I gonna close this bulk PR and work on each one by its own in case I will find time...probably hard to find the proof for you guys if you do not look on the changes itself and think if it is good or bad and can explain why as your use cases are very limited and checking on a UserError in a test is not the best way to test IMHO as it just proves that a condition is triggered not that a result is making sense and here the result of the return will be always no reserved quant whether the quantity is triggering a UserError or not.

**Comment by sle-odoo** [comment]
@wtaferner I'm not starting with a particular usecase, i just say that your branch is red and this test is legit and is broken. Maybe your changes make sense one by one, but together they result in a branch where some functionalities are lost.

The constructive way to work on inventory issues is to present a problematic usecase faced by the user and then look at how to fix the code. So help me help you and describe one.

**Comment by wtaferner** [comment]
@sle-odoo 
I start splitting it to work on them in an isolated fashion...can you in the meantime rethink and for the most explain why it is important to check on an UserError instead of the returned result? I removed the blocker in case there is no quant for this as I was stuck during a normal inventory adjustment for some reason and it was on a product which was perfectly fine IMHO.

**Comment by sle-odoo** [comment]
@wtaferner It's just the way it's designed, if you try to reserve more than what's possible, an exception occurs. You can argue that we should be raising an UserError instead of a more specific exception though. If you reserve the available quantity or less, a list of tuple is returned with the quantity and the quants on which it has been reserved. That's the flow.

**Comment by wtaferner** [comment]
@sle-odoo 
I really hope that we can find a constructive way to resolve this matter based on point of views. I am perfectly ok with a reasonable explanation why it was done like this for every PR I did now. But asking for use cases and not knowing the own use cases and why it is important to be like that is difficult even though I understand that it is not easy and time consuming (for me too by the way). 

The way it is designed is also an argument which is not satisfying as we all try to solve issues and problems. Anyway, maybe we can concentrate on the other PR first if you know a better answer than it is like it is and it is breaking a self-fulfilling test in a nutshell.

Here the PRs to take over this mixed stuff which is indeed difficult to tackle in one thread and mixed testing.

https://github.com/odoo/odoo/pull/21913
https://github.com/odoo/odoo/pull/21914
https://github.com/odoo/odoo/pull/21915
https://github.com/odoo/odoo/pull/21916
https://github.com/odoo/odoo/pull/21917

**Pull request closed without merging**